colossalai/pipeline/rpc/_pipeline_base.py

```diff
@@ -360,11 +363,15 @@ def subscribe_producer(self, microbatch_id: int, forward_only: bool):
 
         work_item_from_producer = WorkItem(stage_id, Phase.FORWARD, subscribe_forward_futures, {}, output,
                                         microbatch_id, None, self.num_microbatches, forward_only)
-
-        # add work_item to work_list
+                
+        return work_item_from_producer
+    
+    # TODO(jiangziyue) Profile the side effect of the lock for lifecycle protection and consider a better one.
+    def subscribe_producer(self, microbatch_id: int, forward_only: bool):
+        key = UniqueKey(microbatch_id, Phase.FORWARD)
         with self.work_list_condition_lock:
-            key = UniqueKey(microbatch_id, Phase.FORWARD)
             if key not in self.work_list:
+                work_item_from_producer = self._subscribe_producer(microbatch_id, forward_only)
```

Comment:
For the maintainability, a more detail comment is expected here to illustrate why we put these actions under the lock context.

Reply:
Done.